README.md
# StreamBop
StreamBop is a software that allows you to play music on your livestreams, credit currently playing music, avoid legal issues, allow audio mixing, customizability and more..

# Downloading
Linux users can use Snap.
[![Snap](https://snapcraft.io/streambop/badge.svg)](https://snapcraft.io/streambop)
```
snap install streambop
```

Windows users can check the releases tab, or:

```
winget install streambop
```
on windows 11.


[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fsnaildos%2FStreamBop.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Fsnaildos%2FStreamBop?ref=badge_large)


CHANGELOG.md
# Changelog

<a name="1.6.1"></a>
## 1.6.1 (2023-07-06)

### Miscellaneous

-  BUMP: Version [[2610356](https://github.com/snaildos/StreamBop/commit/2610356410bf315d2fa646f7859ac53e86b402ad)]
-  FIX: Winget [[3192d96](https://github.com/snaildos/StreamBop/commit/3192d96ce65f63de4e8a711b0b106bfcdfae60dc)]
-  FIX: Invalid CSS [[808f05a](https://github.com/snaildos/StreamBop/commit/808f05a901c43fb0de89c97160b3fb2a417b0bfd)]


<a name="1.6.0"></a>
## 1.6.0 (2023-07-06)

### Miscellaneous

-  FIX: Issue regarding titlebar compatibility [[3f1efe4](https://github.com/snaildos/StreamBop/commit/3f1efe49971f36c193fef3e2b7842e4c915d894d)]
-  BUMP: Cache [[42c1821](https://github.com/snaildos/StreamBop/commit/42c1821f4b8823aad3bde319591b0354acdf301c)]
-  Merge branch &#x27;main&#x27; of https://github.com/snaildos/StreamBop [[1cb836e](https://github.com/snaildos/StreamBop/commit/1cb836e93e82c51e8b92c457211461705d9c5daf)]
-  FEAT: Pre-release tasks [[c2a7235](https://github.com/snaildos/StreamBop/commit/c2a7235ff4fd999bb10ea5617c879f456978d6b0)]
-  Merge pull request [#100](https://github.com/snaildos/StreamBop/issues/100) from snaildos/renovate/electron-25.x [[1ff77d6](https://github.com/snaildos/StreamBop/commit/1ff77d65ba8c874ceb8649eb96baeee231020ca7)]
-  Merge pull request [#95](https://github.com/snaildos/StreamBop/issues/95) from snaildos/renovate/axios-1.x [[b310198](https://github.com/snaildos/StreamBop/commit/b310198ccff2124c19c5cce8578fb910161bbeac)]
-  Merge pull request [#99](https://github.com/snaildos/StreamBop/issues/99) from snaildos/renovate/jquery-3.x [[eecb9f0](https://github.com/snaildos/StreamBop/commit/eecb9f0bc889c9ef01b7a61731bbc347438f6329)]
-  Merge pull request [#101](https://github.com/snaildos/StreamBop/issues/101) from snaildos/renovate/electron-builder-24.x [[54e3889](https://github.com/snaildos/StreamBop/commit/54e38894104f37e5e2a256496fd40744983c1bfe)]
-  Merge pull request [#102](https://github.com/snaildos/StreamBop/issues/102) from snaildos/renovate/electron-updater-6.x [[2485bc8](https://github.com/snaildos/StreamBop/commit/2485bc8098bcf48a5816cc3ac23f8985b112fcbd)]
-  Merge pull request [#93](https://github.com/snaildos/StreamBop/issues/93) from snaildos/development [[722b01e](https://github.com/snaildos/StreamBop/commit/722b01e31ba25c252bd679e42d4371349129c44b)]
-  UPDATE: Cache [[e770541](https://github.com/snaildos/StreamBop/commit/e7705419fc3a14f491a76d1e2026cdc6a5edfc99)]
-  FIX: Version Number [[dcf29e4](https://github.com/snaildos/StreamBop/commit/dcf29e4fd053bd466ceecfd23fa76b7880da39da)]
-  FEAT: Winget [[af87e36](https://github.com/snaildos/StreamBop/commit/af87e369f8c38420260d0b93bddbe79b06f6bd58)]
-  FIX: Compatibility [[0e5bb5d](https://github.com/snaildos/StreamBop/commit/0e5bb5db57dcc60f249c31b120be3dd76e76b515)]
-  FEAT: Yarn Berry [[4d48f36](https://github.com/snaildos/StreamBop/commit/4d48f36d229e3fa2d84cfbcd8a165a8feeefbb6b)]
-  Update dependency electron to v25 [[f160f75](https://github.com/snaildos/StreamBop/commit/f160f75a30778ad10fa5c607de195340cff23f94)]
-  Update dependency electron-updater to v6 [[bf63ed3](https://github.com/snaildos/StreamBop/commit/bf63ed36db1c1d686aaca76d5cf199e0f45245b4)]
-  Update dependency electron-builder to v24 [[719531f](https://github.com/snaildos/StreamBop/commit/719531fa278187e883b3a8cbc4fc3bc74782925b)]
-  Update dependency jquery to v3.7.0 [[62d324b](https://github.com/snaildos/StreamBop/commit/62d324b41e9bc8bf0ba55be186cdc8ec0cd7706f)]
-  Update dependency axios to v1.4.0 [[b538b3c](https://github.com/snaildos/StreamBop/commit/b538b3cc8a91ec7e692ce40c6ab52e678002b116)]
-  Rewrite the fader function [[d66260d](https://github.com/snaildos/StreamBop/commit/d66260d56ef8817ead68f368de9f34a1cce447d5)]
-  Standardize the config page [[ac24cb3](https://github.com/snaildos/StreamBop/commit/ac24cb342d5a66008c2dee286382611fa82d3cf3)]
-  Update package.json [[b9a2101](https://github.com/snaildos/StreamBop/commit/b9a2101e75f8ad4fb74c29092a03f3b3f5b3b44d)]
-  Merge pull request [#94](https://github.com/snaildos/StreamBop/issues/94) from snaildos/renovate/electron-23.x [[71a278e](https://github.com/snaildos/StreamBop/commit/71a278ec287b27cbaffb1b19a6a9b9bf5b269ffb)]
-  Update dependency electron to v23.1.4 [[ac66e50](https://github.com/snaildos/StreamBop/commit/ac66e501f0f9511dabf588b7f7949e5a7bd44449)]
-  Rewrite Entire Code Structure [[032c627](https://github.com/snaildos/StreamBop/commit/032c6277da466f6eb7ecb2b36dd1649aef1c4b92)]
-  Update package.json [[391ae33](https://github.com/snaildos/StreamBop/commit/391ae33ead0c108915f9d2ffc5ead581bc52390d)]
-  Transparency &amp; Frame Fixes [[2bffd9f](https://github.com/snaildos/StreamBop/commit/2bffd9f689d5e83e3c7bd6aef1656f22b6134729)]
-  Rewrite all 1-2 year old code. [[f1cb988](https://github.com/snaildos/StreamBop/commit/f1cb988883664a459926e16e05f7d5d30870c893)]
-  Rewrite Blur Module [[b967e35](https://github.com/snaildos/StreamBop/commit/b967e3535289e7a503f488aa2969f396f70ff181)]
-  Merge pull request [#92](https://github.com/snaildos/StreamBop/issues/92) from snaildos/renovate/axios-1.x [[1c49e14](https://github.com/snaildos/StreamBop/commit/1c49e148c29ba34becdd9d2a85844a93dcffe753)]
-  Merge pull request [#90](https://github.com/snaildos/StreamBop/issues/90) from snaildos/renovate/electron-23.x [[bd33517](https://github.com/snaildos/StreamBop/commit/bd3351735aea17d4d2853e9abe5712b572d75e68)]
-  Merge pull request [#91](https://github.com/snaildos/StreamBop/issues/91) from snaildos/renovate/utf-8-validate-6.x [[5421c5a](https://github.com/snaildos/StreamBop/commit/5421c5a11e507cd1debcd40bdce27edd1cd3c93b)]
-  Merge pull request [#89](https://github.com/snaildos/StreamBop/issues/89) from snaildos/renovate/jquery-3.x [[c114004](https://github.com/snaildos/StreamBop/commit/c114004d26972c9612f595e638b2846e175c62e6)]
-  Update dependency axios to v1.3.4 [[aeae588](https://github.com/snaildos/StreamBop/commit/aeae588d66a344a4025e923c2af9207d625e37d1)]

lib/analytics.js
const ua = require('universal-analytics');
const { uuid } = require('uuidv4');
const Store = require('electron-store');
const tracker = new Store();

// Retrieve the userid value, and if it's not there, assign it a new uuid.
const userId = tracker.get('userid') || uuid();

// (re)save the userid, so it persists for the next app session.
tracker.set('userid', userId);

const usr = ua('UA-156049282-1', userId);

function trackEvent(category, action, label, value) {
if (tracker.get("analytics")) {
console.log("Event triggered to send")
  usr
    .event({
      ec: category,
      ea: action,
      el: label,
      ev: value,
    })
    .send();
console.log("Event sent to Google")
} else if ((tracker.get("analytics")) == false) {
console.log("Google Analytics is disabled.")
} else {
tracker.set("analytics", false);
}}

module.exports = { trackEvent };

main.js
const { autoUpdater } = require('electron-updater');
const { SSL_OP_EPHEMERAL_RSA } = require('constants');
const { app, BrowserWindow, ipcMain } = require('electron');
const { trackEvent } = require('./lib/analytics.js');
const {TitlebarRespect} = require('electron-titlebar-respect')
const glasstron = require('glasstron-clarity');
const electron = require('electron');
electron.app.commandLine.appendSwitch("enable-transparent-visuals");
TitlebarRespect({})

// is dev
var isDev = require('isdev')
// Notify
const { Notification } = require('electron')

// Start the libaries
require('./lib/rpc.js');
console.log("RPC lib init.");

// Prepare database
const Store = require('electron-store');
const config = new Store();

console.log("Software ready for start!");
require('./lib/core.js');


function die() {
console.log("Unregistering variables...");
config.set('newDiff', "unloaded")
config.set('oldDiff', "unloaded")
mainWindow.removeAllListeners('closed');
strm.removeAllListeners('closed');
strm.close()
app.quit()
}

function strmex() {
console.log("Unregistering variables...");
config.set('newDiff', "unloaded")
config.set('oldDiff', "unloaded")
app.removeAllListeners('window-all-closed');
mainWindow.removeAllListeners('closed');
mainWindow.close()
app.quit()
}

function neterr() {
  const notification = {
    title: 'StreamBop',
    body: 'No valid network connection! Please reconnect!'
  }
  new Notification(notification).show()
}

// wait function
function wait(ms)
{
    var d = new Date();
    var d2 = null;
    do { d2 = new Date(); }
    while(d2-d < ms);
}

// Start the libaries
require('./lib/rpc.js');
console.log("RPC lib init.");


// Loading screen
/// create a global var, wich will keep a reference to out loadingScreen window
let loadingScreen;
const createLoadingScreen = () => {
  loadingScreen = new BrowserWindow(
    Object.assign({
      width: 800,
      height: 800,
      alwaysOnTop: true,
      frame: false,
      fullscreen: false,
      show: false
    })
  );
  loadingScreen.setResizable(false);
  loadingScreen.loadFile('./src/pages/splash.html');
  loadingScreen.on('closed', () => (loadingScreen = null));
  loadingScreen.hide();
  loadingScreen.show()
  setTimeout(() => {
    console.log("Main WIN.REQ has been sent now.")
    createWindow()
  }, 2500);
};
console.log("Loading screen ready.");

// Start the main program
let mainWindow;

function createWindow() {
  mainWindow = new glasstron.BrowserWindow({
    width: 900,
    height: 700,
    show: false,
    fullscreen: false,
    frame: global.frame,
    titleBarStyle: global.titleBarStyle,
    modal: true,
    blur: true,
    blurType: global.blurType,
    icon: 'snailfm.ico',
    webPreferences: {
      nodeIntegration: true,
      devTools: isDev,
      nativeWindowOpen: true,
      backgroundThrottling: false,
      contextIsolation: false
    },
  });
  mainWindow.setResizable(false);
  console.log("We are going to send a google analytics streambop.started event")
  trackEvent('streambop.started', 'StreamBop sucessfully started!');
  console.log("We just sent the event. Now let's continue loading")
  mainWindow.setMenuBarVisibility(false)
  mainWindow.loadFile('index.html');
  mainWindow.on('maximize', () => mainWindow.unmaximize());
  wait(6000)
  console.log("Loaded sucess.")
  loadingScreen.close();
  mainWindow.on('closed', () => (
  die()
  ));
// Don't tell me my code is messy, I know it is.
if(isDev) {
  console.log("In Development!")
} else {
  console.log("Not in Development!")
}
// This is the final part of the StreamBop boot-up.
var internetAvailable = require("internet-available");
// Set a timeout and a limit of attempts to check for connection
internetAvailable({
    timeout: 2000,
    retries: 3,
}).then(function(){
    console.log("Internet available");
}).catch(function(){
    console.log("No internet");
    neterr()
    mainWindow.loadFile('./src/pages/nonet.html');
});
    mainWindow.show();
    console.log("Ok! Window init, let's check for updates...")
    autoUpdater.checkForUpdatesAndNotify();
    console.log("Checking for updates... Waiting for a reply back...");
  }

// After Preload
console.log("Main screen ready.");

// strm
/// create a global var, wich will keep a reference to out loadingScreen window
let strm;

const createNewstrm = () => {
  strm = new BrowserWindow(
    Object.assign({
      width: 800,
      setResizable: false,
      isClosable: false,
      minimizable: false,
      height: 600,
      alwaysOnTop: false,
      frame: true,
      fullscreen: false,
      show: false,
      frame: global.frame,
      titleBarStyle: global.titleBarStyle,
      trafficLightPosition: { x: 24, y: 25 },
      webPreferences: {
        nodeIntegration: true,
        devTools: isDev,
        nativeWindowOpen: true,
        backgroundThrottling: false,
        contextIsolation: false
      },
    })
  );
  strm.setResizable(false);
  strm.loadFile('./src/pages/strm.html');
  strm.isMovable(false)
  strm.setMenuBarVisibility(false)
  strm.on('closed', () => (strmex()));
};


app.on('ready', () => {
console.log("strm red");
config.set("on", false);
createNewstrm();
})


app.on('ready', () => {
  createLoadingScreen();
  console.log("Send check for updates signal...");
  console.log("Alright, lets go!");
})


app.on('window-all-closed', function () {
  die()
  if (process.platform !== 'darwin') {
    die()
    app.quit();
  }
});

ipcMain.on('app_version', (event) => {
  event.sender.send('app_version', { version: app.getVersion() });
});

autoUpdater.on('update-available', () => {
  mainWindow.webContents.send('update_available');
});

autoUpdater.on('update-downloaded', () => {
  mainWindow.webContents.send('update_downloaded');
});

ipcMain.on('relaunch', () => {
  mainWindow.close()
});

ipcMain.on('restart_app', () => {
  autoUpdater.quitAndInstall();
});

ipcMain.on('strm', () => {
  strm.show()
});

ipcMain.on('strm.hide', () => {
  strm.hide()
});

ipcMain.on('strm.start', () => {
  console.log("StreamBop viewer activated")
  config.set("on", true);
});

ipcMain.on('strm.stop', () => {
  console.log("StreamBop viewer off")
  config.set("on", false);
});

ipcMain.on('strm.setup', () => {
  console.log("StreamBop viewer setup mode")
  config.set("on", false);
  config.set("setupan", true);
});

ipcMain.on('strm.reload', () => {
  strm.reload();
});

ipcMain.on('minimize', () => {mainWindow.minimize()})
ipcMain.on('restore', () => {mainWindow.restore()})
ipcMain.on('close', () => {mainWindow.close()})